napari/_qt/dialogs/_tests/test_screenshot_dialog.py

```diff
@@ -0,0 +1,89 @@
+import pytest
+from qtpy.QtCore import QTimer
+from qtpy.QtWidgets import QApplication, QFileDialog, QLineEdit, QMessageBox
+
+from napari._qt.dialogs.screenshot_dialog import ScreenshotDialog
+from napari.utils.history import get_save_history
+
+
+@pytest.mark.parametrize("filename", ["test", "test.png", "test.tif"])
+def test_screenshot_save(qtbot, tmp_path, filename):
+    """Check passing different extensions with the filename."""
+
+    def save_function(path):
+        # check incoming path has extension event when a filename without one
+        # was provided
+        assert filename in path
+        if "." not in filename:
+            assert ".png" in path
+
+        # create a file with the given path to check for
+        # non-native qt overwrite message
+        with open(path, "w") as mock_img:
+            mock_img.write("")
+
+    qt_overwrite_shown = False
+
+    def handle_qt_overwrite_message():
+        for widget in QApplication.topLevelWidgets():
+            if isinstance(widget, QMessageBox):
+                # test should not enter here!
+                widget.accept()
+                nonlocal qt_overwrite_shown
+                qt_overwrite_shown = True
+                break
```

Comment:
I ended up with this logic since I tried patching the `warning` method as in the link but the patch was not being triggered. Maybe patching `exec` could work 🤔. My initial guess was that since [the warning `QMessageBox` is instantiated in Qt code (C++)](https://code.qt.io/cgit/qt/qtbase.git/tree/src/widgets/dialogs/qfiledialog.cpp?h=dev#n2669) the monkey patching didn't have any effect on it but I will check!

Reply:
Checked and seems like even when monkey patching `exec` the Qt generated `QMessageBox` warning message keeps appearing.

If I do the monkey patch and revert the changes, to first call `super().accept` and then the `save_function` in the `accept` method for the dialog, the test hangs until I manually close the Qt generated `QMessageBox` warning:

![imagen](https://github.com/napari/napari/assets/16781833/fefd2065-61d1-4713-966d-38d4d18ba024)

Maybe I'm missing something in the monkey patch procedure @Czaki ? 🤔 

```python
@pytest.mark.parametrize("filename", ["test", "test.png", "test.tif"])
def test_screenshot_save(qtbot, tmp_path, monkeypatch, filename):
    """Check passing different extensions with the filename."""

    def save_function(path):
        # check incoming path has extension event when a filename without one
        # was provided
        assert filename in path
        assert "." in filename or ".png" in path

        # create a file with the given path to check for
        # non-native qt overwrite message
        with open(path, "w") as mock_img:
            mock_img.write("")

    def qt_overwrite_qmessagebox_warning(*_, **__):
        raise RuntimeError("exec_ call")

    monkeypatch.setattr(QMessageBox, "exec_", qt_overwrite_qmessagebox_warning)
    monkeypatch.setattr(QMessageBox, "critical", qt_overwrite_qmessagebox_warning)
    monkeypatch.setattr(QMessageBox, "information", qt_overwrite_qmessagebox_warning)
    monkeypatch.setattr(QMessageBox, "question", qt_overwrite_qmessagebox_warning)
    monkeypatch.setattr(QMessageBox, "warning", qt_overwrite_qmessagebox_warning)

    # setup dialog
    dialog = ScreenshotDialog(
        save_function, directory=str(tmp_path), history=[]
    )
    qtbot.addWidget(dialog)
    dialog.setOptions(QFileDialog.DontUseNativeDialog)
    dialog.show()

    # check dialog and set filename
    assert dialog.windowTitle() == 'Save screenshot'
    line_edit = dialog.findChild(QLineEdit)
    line_edit.setText(filename)

    # check that no warning message related with overwriting is shown
    dialog.accept()

    # check the file was created
    save_filename = filename if '.' in filename else f'{filename}.png'
    qtbot.waitUntil(lambda: (tmp_path / save_filename).exists())
```

Reply:
Oh. So it looks like Your procedure was better as maybe Qt uses inner references.